### ISSUE TITLE
Support for Audio-Reactive Lighting Systems (WLED/LEDfx) via DMX/sACN/rUDP

### ISSUE THREAD

**Issue by qu4nt4r** [raveberry/raveberry#69]
Hi there, 
as I am one of many who are building up LED systems via [WLED](https://github.com/Aircoookie/WLED), raveberry would be a perfect fit for a mobile light-and-sound system. Would... With my limited coding knowledge I assume there currently is no natural solution to combine raveberry with such esp/ws2812b-based lighting systems. 
Would be awesome having (outdoor) parties where everyone can contribute to such a audio reactive experience.
Greeting
KeLvin

**Comment by raveberry** [comment]
Hi,
currently, Raveberry is already able to control leds. In comparison to WLED, the visualization options are pretty limited, there is an audio reactive program and some audio independent ones.
I use a WS281x NeoPixel ring and a single-colored strip controlled by an Adafruit PWM module. It should be pretty simple to add a third output alternative for WLED. The simplest option which would easily fit into the current design would be to use the [realtime interface](https://github.com/Aircoookie/WLED/wiki/UDP-Realtime-Control).
In order to make use of WLED's patterns, more work would be required, but it should still be possible.
How did you imagine this new feature to work with Raveberry?

**Comment by 123vollhorst** [comment]
Hi, thanks for the fast response.
Pardon me, I think I had a little thinking mistake (things get complicated with time ;). 
Yes, [LEDfx](https://github.com/ahodges9/LedFx) can run on the RavePi and utilize a virtual input by use of the sound output. With WLED the things are a bit different as it doesnt work on a Pi but mostly on esp32´s which are normally distributed within the wifi network. But as a workaround, the line-out of the RavePi can be connected to the [line-in of a sound reactive esp32](https://github.com/atuline/WLED/wiki/Audio-Input-Options) (assuming there is a line-in option like with one of the new [esp32 based audio boards](https://github.com/espressif/esp-adf)) that handles FFT and sends the signal to other esp via a newly integrated UDP-Audio-Sync function. 
That brings me to a more concrete question: Is there a way to output music on a RavePi via line-out _and_ bluetooth? As the sound output to the decoding esp is a dead end (i.e. no throughput), streaming the sound via bluetooth to a speaker would give music to both; to the audience to hear and to the esp to decode for sound reactive lighting.

**Comment by raveberry** [comment]
Yes it should be possible possible to output to multiple devices. I think there is a pulse module that broadcasts to all other sinks.

The question is if we want the sound reactive esp to perform the FFT. As I said, Raveberry already does this. It uses [cava](https://github.com/karlstav/cava) to visualize the frequencies that are currently playing.

As I see it there are two options:
1. Use the realtime interface directly from Raveberry. This would reuse the code already computing led visualization and provide a uniform display across all led outputs. However, WLED's additional programs and functions would not be accessible, although they could be implemented by using the http api.
2. Have an additional device do the sound transformation. Then we could use more of WLED's mature functions. There are some drawbacks, though: As you mentioned, the audio jack would be occupied by the connecting cable. Also, such a reactive esp would be required.

So my suggestion would be to add WLED's realtime interface as a third output for Raveberry (in addition to WS281x and PWM). Maybe you'd have to supply your number of leds and the ip's of the esp's controlling these leds. Then Raveberry's light programs (static color, rainbow fade, flash and audio-reactive) would be available. The relevant code is in [this folder](https://github.com/raveberry/raveberry/tree/master/core/lights).
As a next step we could extend the interface to include WLED's commands.

What do you think about that?

**Comment by minz1** [comment]
There's actually a fork of WLED that has audio reactive stuff, over here -> https://github.com/atuline/WLED

**Comment by raveberry** [comment]
Raveberry 0.7.0 now features WLED support. It uses the UDP realtime interface to control them.
At `/lights` there is now a "WLED" section where you can specify an IP and the number of LEDs of your strip. Setting the program to "Rave" provides audio reactive lighting.

I'd be happy if you try it out and tell me how it went.

**Comment by 123vollhorst** [comment]
Thanks! Will try it out asap. I love the open source community. 
Greetings
Kelvin

**Comment by champagj** [comment]
I have done a couple projects on ESP using smart leds. I have an esp8266 and esp32 I can use to test and I just ordered some leds. Never used WLED but I will give it a try.

**Comment by champagj** [comment]
I installed WLED built 2007190 (main fork from Aircookie)  on an ESP8266 compiled on Arduino IDE (a few libraries needed to download..) using 8212 strip.

Config Raveberry with IP of WLED in /lights section. Upon connect the WLED GUI shows this message.

![image](https://user-images.githubusercontent.com/47355785/89368477-a5efe180-d6a9-11ea-8633-add2ba9eba86.png)

Once you override raveberry is able to control the leds. 
The fixed color option works as expected. The color rendering appears a bit off but the main colors red, blue etc are okay
The Rainbow and speed options works as well.

The Rave option does not appear to do anything..  it keeps the previous effect.

Is this suppose to be sound-reactive? I did not use the WLED fork from atuline that is dedicated to soud-reactive WLED.

**Comment by champagj** [comment]
Further testing of /lights
I went to turn lights back on as raveberry was switching to the next song

![image](https://user-images.githubusercontent.com/47355785/89370466-ccfce200-d6ae-11ea-91c6-f607667e17da.png)

Mopidy is in fact playing the next song but not displayed by raveberry. 
If I tried to play next song, mopidy stopsplayback and raveberry does not cue next song and stops.

I did a reboot server and the next cued song moved to current and started playing.

**Comment by raveberry** [comment]
The main fork is sufficient, Raveberry does the sound-reactive part on its own.

Your Raveberry probably crashed, because you tried to enable the Rave program without the dependencies being installed. This also explains the weird behavior when switching the song. I should hide this option in this case. 

Raveberry needs to install a program that performs the fft on captured sound. As this requires some system configuration, it is disabled by default. In order to activate it, you need to set `led_visualization = True` in your .ini file during Installation. You can reinstall your Raveberry without any data loss.

The Fixed and Rainbow programs work without this additional dependency. Great to hear that the WLED do not only work in my setup :D

**Comment by champagj** [comment]
Thanks for the explanation, I will do the update to install the led_visualisation to test.

I would suggest to hide/gray out the entire WLED panel if the led_visualisation is not installed to match the behavior for the other panels/tabs under /lights.

In my own opinion the interface to WLED needs to provide audio-reactive to make sense.

**Comment by champagj** [comment]
I ran  the raveberry full-install with led_visualisation =true.
Raveberry was working prior but running the installer caused issues.
Trying to load a song and they all appear briefly in the current song window and disappear.

  GNU nano 3.2                                             info.log

2020-08-05 13:40:15 wsclient.__init__:36 INFO  Creating Mopidy Websocket connection...
2020-08-05 13:40:19 wsclient.__init__:36 INFO  Creating Mopidy Websocket connection...
2020-08-05 13:40:22 wsclient.__init__:36 INFO  Creating Mopidy Websocket connection...
2020-08-05 13:40:31 cli.run:287 INFO  Starting server at tcp:port=9000:interface=0.0.0.0
2020-08-05 13:40:31 server.run:111 INFO  HTTP/2 support not enabled (install the http2 and tls Twisted extras)
2020-08-05 13:40:31 server.run:119 INFO  Configuring endpoint tcp:port=9000:interface=0.0.0.0
2020-08-05 13:40:31 server.listen_success:153 INFO  Listening on TCP address 0.0.0.0:9000
2020-08-05 13:40:31 base.decode:114 WARNING  Session data corrupted
2020-08-05 13:40:33 wsclient.__init__:36 INFO  Creating Mopidy Websocket connection...
2020-08-05 13:40:46 server.kill_all_applications:310 INFO  Killed 1 pending application instances
2020-08-05 13:41:08 cli.run:287 INFO  Starting server at tcp:port=9000:interface=0.0.0.0
2020-08-05 13:41:08 server.run:111 INFO  HTTP/2 support not enabled (install the http2 and tls Twisted extras)
2020-08-05 13:41:08 server.run:119 INFO  Configuring endpoint tcp:port=9000:interface=0.0.0.0
2020-08-05 13:41:08 server.listen_success:153 INFO  Listening on TCP address 0.0.0.0:9000
2020-08-05 13:41:09 base.decode:114 WARNING  Session data corrupted
2020-08-05 13:41:12 wsclient.__init__:36 INFO  Creating Mopidy Websocket connection...
2020-08-05 13:42:36 base.decode:114 WARNING  Session data corrupted
2020-08-05 13:42:36 base.decode:114 WARNING  Session data corrupted
2020-08-05 13:42:43 base.decode:114 WARNING  Session data corrupted


FileNotFoundError: [Errno 2] No such file or directory: 'cava': 'cava'
2020-08-05 13:52:52 log.log_response:228 ERROR  Internal Server Error: /ajax/settings/reboot_system/
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/dist-packages/django/core/handlers/exception.py", line 34, in inner
    response = get_response(request)
  File "/usr/local/lib/python3.7/dist-packages/django/core/handlers/base.py", line 115, in _get_response
    response = self.process_exception_by_middleware(e, request)
  File "/usr/local/lib/python3.7/dist-packages/django/core/handlers/base.py", line 113, in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "./core/settings/settings.py", line 63, in _decorator
    self.base.settings.update_state()
  File "./core/state_handler.py", line 35, in update_state
    send_state_event(self.state_dict())
  File "./core/state_handler.py", line 18, in send_state_event
    async_to_sync(channel_layer.group_send)("state", data)
  File "/usr/local/lib/python3.7/dist-packages/asgiref/sync.py", line 114, in __call__
    self._run_event_loop, loop, awaitable
  File "/usr/lib/python3.7/concurrent/futures/thread.py", line 153, in submit
    raise RuntimeError('cannot schedule new futures after '
RuntimeError: cannot schedule new futures after interpreter shutdown

**Comment by raveberry** [comment]
Reinstalling Raveberry also resets the mopidy config. So unfortunately you need to set the credentials again.

Thanks for the error message during install. Apparently I never tested the installation on a system where git was not installed...
I will fix this in the next update. Until then you can install it manually by running these commands:
```
sudo su
cd /opt
apt-get install -y git libfftw3-dev libasound2-dev libncursesw5-dev libpulse-dev libtool m4 automake libtool
git clone https://github.com/karlstav/cava
cd cava
./autogen.sh
./configure
make
make install
```

Another reinstall should not be necessary.

**Comment by minz1** [comment]
For some reason, the Rave effect doesn't seem to work for me. The others work, but it gets reset and then stuck on whatever effect I have on when I switch to Rave.

**Comment by raveberry** [comment]
Probably you have the same problem and cava did not install correctly. I just released 0.7.1, which installs the git dependency correctly. So simply upgrading your installation should do the trick.
